### PR TITLE
Crafting input interface terminal supports

### PR DESCRIPTION
**Pull request by harrynull**
* Make crafting input hatches show up in interface terminal.
* Custom name with quartz knife
* Block highlight in crafting status
* Does not support the wireless versions

Other half in https://github.com/GTNewHorizons/GT5-Unofficial/pull/2200 which needs to be merged after.

Closes https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/14116
Closes https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/14088

### CODE REVIEW

**Review by firenoo — CHANGES_REQUESTED** [review]
Shouldn't make a hard dep on GT5

**Comment by miozune** [comment]
Rather than having hardcoded support for specific tile, adding API for addons would be better. We have large molecular assembler for potential support as well.
Example impl: Add interface (`IInterfaceTerminalSupport` or whatever you name), and make `IInterfaceHost` and GT tile to implement it. It can have some methods to provide `InvTracker` and `invoke` behavior. Add new registry to `IRegistryContainer` and accept `Class<? extends IInterfaceTerminalSupport>` to filter by `getMachines`. Add `getTile` to `MetaTileEntity` with comment explaining it's accessed via reflection.

**Comment by harrynull** [comment]
@firenoo @miozune 

refactored to remove dependency to gt, and instead exposed an api. would appreciate another look

File: src/main/java/appeng/helpers/IInterfaceTerminalSupport.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
+package appeng.helpers;
```
Comment (firenoo) [inline src/main/java/appeng/helpers/IInterfaceTerminalSupport.java:1]: I am not too confident about this. Despite miozune's suggestion, I think it is more appropriate to reuse an existing interface, such as `ICraftingProvider` which has the properties we want.

File: src/main/java/appeng/container/implementations/ContainerInterfaceTerminal.java [hunk 2]
```diff
@@ -92,22 +88,20 @@ public void detectAndSendChanges() {
...
-                    total += interfaceCheck.getTotal();
-                    missing |= interfaceCheck.isMissing();
-                }
-
-                for (final IGridNode gn : this.grid.getMachines(PartP2PInterface.class)) {
-                    InterfaceCheck interfaceCheck = new InterfaceCheck().invoke(gn);
-                    total += interfaceCheck.getTotal();
-                    missing |= interfaceCheck.isMissing();
+                var clzWithInterfaceSupport = StreamSupport.stream(grid.getMachinesClasses().spliterator(), false)
+                        .filter(IInterfaceTerminalSupport.class::isAssignableFrom).distinct().toArray(Class[]::new);
```
Comment (firenoo) [inline src/main/java/appeng/container/implementations/ContainerInterfaceTerminal.java:92]: I do not like how it needs to constantly check every single machine type on the Grid every time `detectAndSendChanges` is called and create a new class list every time. Having mods register their classes at init/postInit is better.

On the other hand, a major refactor needs to happen here anyways given the performance complaints that we've received already.

File: src/main/java/appeng/container/implementations/ContainerInterfaceTerminal.java [hunk 3]
```diff
@@ -57,7 +53,7 @@ public final class ContainerInterfaceTerminal extends AEBaseContainer {
      */
     private static long autoBase = Long.MIN_VALUE;
 
-    private final Multimap<IInterfaceHost, InvTracker> diList = HashMultimap.create();
+    private final Multimap<IInterfaceTerminalSupport, InvTracker> diList = HashMultimap.create();
```
Comment (firenoo) [inline src/main/java/appeng/container/implementations/ContainerInterfaceTerminal.java:56]: nit: Needs a different name, was duality interface list but now we've expanded to a more general term (probably crafting providers).

**Review by firenoo — DISMISSED** [review]

File: src/main/java/appeng/helpers/IInterfaceTerminalSupport.java (hunk 1 above)
Comment (eigenraven) [reply src/main/java/appeng/helpers/IInterfaceTerminalSupport.java:1]: As long as all new methods have a `default` implementation, it's backwards compatible

File: src/main/java/appeng/container/implementations/ContainerInterfaceTerminal.java (hunk 2 above)
Comment (harrynull) [reply src/main/java/appeng/container/implementations/ContainerInterfaceTerminal.java:92]: I think that shouldn't be too bad performance-wise? There are not too many machine types and it probably won't be the slowest thing in the code right?

File: src/main/java/appeng/helpers/IInterfaceTerminalSupport.java (hunk 1 above)
Comment (miozune) [reply src/main/java/appeng/helpers/IInterfaceTerminalSupport.java:1]: | Part Name  | ICraftingProvider | Support Interface Terminal | IInterfaceHost |
| - | - | - | - |
| Level Emitter | ○ | × | × |
| Crafting Input Buffer  | ○ | ○ | × |
| ME Interface | ○ | ○ | ○ |

`ICraftingProvider` does not necessarily have ability to support interface terminal. Having separated interface contributes to reducing unnecessary methods declared on `ICraftingProvider`. If needed we can make `IInterfaceTerminalSupport` extend `ICraftingProvider`.

File: src/main/java/appeng/container/implementations/ContainerInterfaceTerminal.java (hunk 2 above)
Comment (firenoo) [reply src/main/java/appeng/container/implementations/ContainerInterfaceTerminal.java:92]: We'll need to rework this after stable

File: src/main/java/appeng/container/implementations/ContainerInterfaceTerminal.java (hunk 3 above)
Comment (harrynull) [reply src/main/java/appeng/container/implementations/ContainerInterfaceTerminal.java]: fixed

File: src/main/java/appeng/container/implementations/ContainerInterfaceTerminal.java (hunk 2 above)
Comment (harrynull) [reply src/main/java/appeng/container/implementations/ContainerInterfaceTerminal.java]: done

**Review by firenoo — CHANGES_REQUESTED** [review]
Sorry for the hassle, we had a discussion on the interface terminal right before this PR came in so I wasn't sure how this would affect those plans

tested a bit and looks fine other than the noted change

File: src/main/java/appeng/container/implementations/ContainerInterfaceTerminal.java [hunk 4]
```diff
@@ -92,30 +87,30 @@ public void detectAndSendChanges() {
...
+                    }
+                    // we can stop if any is missing. The value of `total` is not important if `missing == true`
+                    if (missing) break;
                 }
             }
         }
 
-        if (total != this.diList.size() || missing) {
+        if (total != this.supportedInterfaces.size() || missing) {
+            System.out.println(
```
[the remark this answers is not in the document]
Comment (harrynull) [reply src/main/java/appeng/container/implementations/ContainerInterfaceTerminal.java]: Good catch. Forgot to remove those. Fixed

**Review by firenoo — APPROVED** [review]
~AE2FC will need a patch to get this to work, but that can be done after this gets merged (assuming no rush)~

Thanks to good design, we don't have to worry about this

**Comment by Darkenlord1** [comment]
Any hope for Large molecular assembler patterns interface visibility?

**Pull request merged**